Bottom: 5d156e7750c68c5ec7bab8c8982fdd2d8e2ef90b
Top:    309aaf7f3640df3893921a937ba63460a370cebc
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-17 07:44:14 +0800

tuntap: enable bh early during processing XDP

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 4506da3..7ecc4a0 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1725,22 +1725,18 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 			goto err_xdp;
 		}
 	}
+	rcu_read_unlock();
+	local_bh_enable();
 
 	skb = build_skb(buf, buflen);
-	if (!skb) {
-		rcu_read_unlock();
-		local_bh_enable();
+	if (!skb)
 		return ERR_PTR(-ENOMEM);
-	}
 
 	skb_reserve(skb, pad - delta);
 	skb_put(skb, len);
 	get_page(alloc_frag->page);
 	alloc_frag->offset += buflen;
 
-	rcu_read_unlock();
-	local_bh_enable();
-
 	return skb;
 
 err_redirect:
